environment/src/main/java/com/sequenceiq/environment/environment/poller/StackPollerProvider.java

```diff
@@ -0,0 +1,83 @@
+import com.sequenceiq.environment.environment.service.stack.StackService;
+import com.sequenceiq.flow.core.FlowConstants;
+import com.sequenceiq.flow.domain.FlowLog;
+import com.sequenceiq.flow.service.flowlog.FlowLogDBService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.ws.rs.BadRequestException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class StackPollerProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StackPollerProvider.class);
+
+    private final StackService stackService;
+
+    private final FlowLogDBService flowLogDBService;
+
+    public StackPollerProvider(
+        StackService stackService,
+        FlowLogDBService flowLogDBService) {
+        this.stackService = stackService;
+        this.flowLogDBService = flowLogDBService;
+    }
+
+    public AttemptMaker<Void> stackUpdateConfigPoller(List<String> stackNames, Long envId, String flowId) {
+        List<String> mutableNamesList = new ArrayList<>(stackNames);
+        return () -> {
+            Optional<FlowLog> flowLog = flowLogDBService.getLastFlowLog(flowId);
+            if (flowLog.isPresent() && flowLog.get().getCurrentState().equals(FlowConstants.CANCELLED_STATE)) {
+                return AttemptResults.finishWith(null);
+            }
+            List<String> remaining = new ArrayList<>();
+            List<AttemptResult<Void>> results = collectStackUpdateConfigResults(mutableNamesList,
+                remaining, envId);
+            mutableNamesList.retainAll(remaining);
+            return evaluateResult(results);
+        };
+    }
+
+    private List<AttemptResult<Void>> collectStackUpdateConfigResults(List<String> stackNames,
+        List<String> remaining, Long envId) {
+        return stackNames.stream()
+            .map(stackName -> fetchStackUpdateConfigResults(remaining, stackName, envId))
+            .collect(Collectors.toList());
+    }
+
+    private AttemptResult<Void> fetchStackUpdateConfigResults(List<String> remainingStacks,
+        String stackName, Long envId) {
+        try {
+            stackService.triggerConfigUpdateForStack(stackName);
+            return AttemptResults.finishWith(null);
+        } catch (BadRequestException e) {
+            remainingStacks.add(stackName);
+            return AttemptResults.justContinue();
+        } catch (Exception e) {
+            return AttemptResults.breakFor(e);
+        }
+    }
+
+    AttemptResult<Void> evaluateResult(List<AttemptResult<Void>> results) {
+        Optional<AttemptResult<Void>> error = results.stream()
+            .filter(it -> it.getState() == AttemptState.BREAK).findFirst();
+        if (error.isPresent()) {
+            return error.get();
+        }
+        long count = results.stream().filter(it -> it.getState() == AttemptState.CONTINUE).count();
+        if (count > 0) {
```

Comment:
I think you should use `anyMatch` here instead of `count` and `filter` if I understand the code well

Reply:
Can do it, but it ends up being the same thing.  Not sure which is more clear. 

```
Predicate<AttemptResult<Void>> stateCheck = it -> it.getState() == AttemptState.CONTINUE;
if (results.stream().anyMatch(stateCheck)) {
```
versus

```
long count = results.stream().filter(it -> it.getState() == AttemptState.CONTINUE).count();
if (count > 0) {
```

I think the `anyMatch` adds more overhead needing an additional object, but in the scheme of things  it really makes no difference.

Reply:
yeah, that's not better, but this is what I was trying to suggest:
```
		boolean continueResult = results.stream().anyMatch(result -> result.getState() == AttemptState.CONTINUE);
        if (continueResult) {
            return AttemptResults.justContinue();
        }
```
or even better moving this to a method:
```
        if (shouldContinue(results)) {
            return AttemptResults.justContinue();
        }
...
    private boolean shouldContinue(List<AttemptResult<Void>> results) {
        return results.stream().anyMatch(result -> result.getState() == AttemptState.CONTINUE);
    }
```

Reply:
sounds good to me.